index.js

```diff
@@ -1,17 +1,41 @@
-"use strict";
+'use strict';
 
 // Load rules
 var Trie = require('./lib/suffix-trie.js');
 var allRules = Trie.fromJson(require('./rules.json'));
 
-var cleanHostValue = require('./lib/clean-host.js');
+// Internals
+var extractHostname = require('./lib/clean-host.js');
 var getDomain = require('./lib/domain.js');
 var getPublicSuffix = require('./lib/public-suffix.js');
 var getSubdomain = require('./lib/subdomain.js');
 var isValid = require('./lib/is-valid.js');
 var tldExists = require('./lib/tld-exists.js');
 
 
+function parse(url, validHosts, rules, _extractHostname) {
+  var hostname = _extractHostname(url);
+  var valid = isValid(hostname);
+  var publicSuffix = null;
+  var domain = null;
+  var subdomain = null;
+
+  if (valid) {
+    publicSuffix = getPublicSuffix(rules, hostname);
+    domain = getDomain(validHosts, publicSuffix, hostname);
+    subdomain = getSubdomain(hostname, domain);
+  }
+
+  return {
+    valid,
+    hostname,
```

Comment:
Thinking about it, the only property I would add would be `tldExists`, derived from the same method. Especially as now, thanks to you, we can retrieve a public suffix while having an unknown TLD.

What do you think?

Reply:
Yes we can, this way we actually have an equivalent for each method of the API when using the `parse` function. I will add it.